### PR TITLE
Document: Do not cache panel-zoom tiles to disk and fix their caching and rendering

### LINKED ISSUE #12300: Crash upon closing a file right after panel zoom
* KOReader version: 2024.07-49
* Device: Kobo Forma

#### Steps to reproduce

On a Forma¹, with Panel zoom enabled:

- Open the CBZ provided in https://github.com/koreader/koreader/issues/7961#issue-941562162;
- Tap and hold on the top center panel;
- Close the zoomed-in panel;
- Open the file browser to close the file.

The combination of the panel's small-ish dimensions, its aspect ratio close to the screen's and the Forma's high resolution results in a high zoom factor and a tile size that is apparently too much for our poor little device.

¹ I guess this could also be reproduced on a Sage or Aura One, given their similar specs.

##### `crash.log`

The last lines differ depending on which action was used to close the file, but the beginning remains the same:

```
./luajit: ./ffi/zstd.lua:29: Failed to allocate ZSTD compression buffer (84120300 bytes)
stack traceback:
	[C]: in function 'assert'
	./ffi/zstd.lua:29: in function 'zstd_compress'
	frontend/persist.lua:89: in function 'serialize'
	frontend/persist.lua:244: in function 'save'
	frontend/document/tilecacheitem.lua:45: in function 'dump'
	frontend/document/doccache.lua:97: in function 'serialize'
	frontend/apps/reader/readerui.lua:819: in function 'onClose'
```

### PR DESCRIPTION
**Pull request by NiLuJe**
Because they often be crazy huge, and they were asking DocCache to do crazy, crazy things.

(It also broke DocCache's crappy heuristic that attempts to rewind to the current page to choose which tile to dump to disk, dumping that tile instead of the page)

Fix #12300

----

* Use a dedicated cache hash for partial tiles from panel-zoom
* Never dump them to disk, as it confuses DocCache's crappy heuristics that rewinds the cache to skip over the hinted page to try to dump the on-screen page to disk.
* Apply the zoom factor in the exact same way as any other page rect (i.e., floor coordinates, ceil dimensions), and make sure said rect is actually a Geom so it doesn't break the cache hash, which relies on Geom's custom tostring method for rects. Said scaling method *also* belongs to the Geom class anyway.
* Handle such pre-scaled rects properly in renderPage, so as not to apply the zoom factor to the full page, which would attempt to create a gigantic buffer.
* And now that the rect is rendered properly in an appropriately-sized buffer, use the rendered tile as-is, no need to blit it to another (potentially way too large because of the above issue) blank BB.
* The zoom factor is now computed for a scale to best-fit (honoring `imageviewer_rotate_auto_for_best_fit`), ensuring the best efficiency (ImageViewer won't have to re-scale).
* Cache: Reduce the maximum item size to 50% of the cache, instead of 75%.
* Warn about the legacy ReaderRotation module, as it turned out to be horribly broken. The whole machinery (which is spread over *a lot* of various codepaths) is left as-is, peppered with notes & fixmes hinting at the problem. Thankfully, that's not how we actually handle rotation, so it was probably hardly ever used (which possibly explains why nobody ever noticed it breaking, and that nugget possibly dates back to the inception of the kpv -> ko refactor!). (#12309)

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12303)

### CODE REVIEW

**Comment by NiLuJe** [comment]
Before: https://github.com/koreader/koreader/issues/12300#issuecomment-2270130646

After:

```
MemAvailable:     392916 kB
MemAvailable:     392916 kB
MemAvailable:     392916 kB
MemAvailable:     392900 kB
MemAvailable:     392900 kB
MemAvailable:     392900 kB
MemAvailable:     392900 kB
MemAvailable:     392900 kB
MemAvailable:     392900 kB
MemAvailable:     392900 kB
MemAvailable:     389816 kB
MemAvailable:     388688 kB
MemAvailable:     388564 kB
MemAvailable:     381236 kB
MemAvailable:     381236 kB
MemAvailable:     381236 kB
MemAvailable:     373288 kB
MemAvailable:     371428 kB
MemAvailable:     370792 kB
MemAvailable:     370312 kB
MemAvailable:     369692 kB
MemAvailable:     368936 kB
MemAvailable:     368936 kB
MemAvailable:     368936 kB
MemAvailable:     368936 kB
MemAvailable:     368936 kB
MemAvailable:     368936 kB
MemAvailable:     368936 kB
MemAvailable:     368936 kB
MemAvailable:     368936 kB
MemAvailable:     368936 kB
MemAvailable:     368936 kB
MemAvailable:     368940 kB
MemAvailable:     368940 kB
MemAvailable:     368940 kB
MemAvailable:     371312 kB
MemAvailable:     371188 kB
MemAvailable:     370692 kB
MemAvailable:     370072 kB
MemAvailable:     369452 kB
MemAvailable:     368956 kB
MemAvailable:     368956 kB
MemAvailable:     368956 kB
MemAvailable:     368956 kB
MemAvailable:     368956 kB
MemAvailable:     368956 kB
MemAvailable:     368956 kB
MemAvailable:     368944 kB
MemAvailable:     368944 kB
MemAvailable:     368944 kB
MemAvailable:     368944 kB
MemAvailable:     368944 kB
MemAvailable:     368944 kB
MemAvailable:     380752 kB
MemAvailable:     380752 kB
MemAvailable:     380752 kB
MemAvailable:     380752 kB
MemAvailable:     380752 kB
MemAvailable:     380752 kB
MemAvailable:     380752 kB
MemAvailable:     380752 kB
MemAvailable:     380752 kB
MemAvailable:     380748 kB
MemAvailable:     380748 kB
MemAvailable:     380748 kB
MemAvailable:     380748 kB
```

**Comment by NiLuJe** [comment]
Reviews wanted, because this is a bit out of my wheelhouse, and it's 3AM ;D.

**Comment by NiLuJe** [comment]
(Draft because there's still a bunch of debug printing left)

**Comment by NiLuJe** [comment]
And now I'm mildly wondering if reflow might not be doing something similarly dumb (e.g., #12227)...

**Comment by NiLuJe** [comment]
Quick question @Galunid : what did you mean by "image projection" here?

https://github.com/koreader/koreader/blob/7e8cdbcf657fd2bd914ce41df1994ccdae4c1501/frontend/document/document.lua#L535

(Because I'm *this* close to just passing a sane zoom and letting renderPage" do things right).

**Comment by NiLuJe** [comment]
Refactored to avoid mangling drawPage's already insane signature, and dropped the unnecessary blit in getpagePart( now also renamed drawPagePart, because that's what it does, and matches the other draw* APIs).

**Comment by NiLuJe** [comment]
@fbriere: Since you actually *use* the feature, please feel free to test this extensively ;).

**Comment by Galunid** [comment]
> Quick question @Galunid : what did you mean by "image projection" here?

Those need to be `int` and not `float`, because when you pass down `float` you get image that's transformed in a weird way (stretched, rotated, filled with black triangles in corners). Similar to image projection (input image * transformation matrix -> image projection).

**Comment by NiLuJe** [comment]
Seeing the discussion in https://github.com/koreader/koreader/pull/6709, https://github.com/koreader/koreader/pull/12303/commits/dba95b40776e231b50be8bf8c7038481b9ea972f miiiight need to be a user-setting.

I feel like this covers the vast majority of sensible use-cases, and is vastly more efficient, but, again, not actually a user myself ;).

**Comment by NiLuJe** [comment]
> Those need to be `int`

They still are ;).

Not seeing anything weird so far on my end.

**Comment by NiLuJe** [comment]
> [dba95b4](https://github.com/koreader/koreader/commit/dba95b40776e231b50be8bf8c7038481b9ea972f) miiiight need to be a user-setting.

Or automagic something like computing against the matching dimension and force-feeding a rotation to ImageViewer?

**Comment by Galunid** [comment]
>They still are ;). Dunno if the original implementation predates the Geom tweaks of scaleBy, prehaps explaining the worries at the time.
>
>Not seeing anything weird so far on my end now in any case ;).

Probably, there were kopt and leptonica bumps in the meantime and some other tweaks. Lots of things changed over almost 4 years. Or I didn't know I could've been using `Geom`, which is another option ;)

**Comment by NiLuJe** [comment]
> Or automagic something like computing against the matching dimension and force-feeding a rotation to ImageViewer?

Did that in 51fb5a7, piggy-backing on `imageviewer_rotate_auto_for_best_fit` ;).

**Comment by NiLuJe** [comment]
> From my meager experience with it, I can confirm that it does seem to be less stupidly slow now.

Yup, especially noticeable on the Elipsa, where I only have 1:1 perfect-fit files. I basically only used the feature by accident when the Elan panel spazzes out and trips a hold, and that always lead to a serious stall. Now It Just Works(TM). Yay.

**Comment by NiLuJe** [comment]
Turns out the legacy rotation handling that's taking up so much code all over the place is broken, lol.

Anyway, left the mess in place, just disabled ReaderRotation, because if you ever were to hit the hilariously aptly named key binds: J/K, ytou'd end up with only blank pages.

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 3 of 5 files at r4, 1 of 2 files at r7, 5 of 6 files at r8, 1 of 1 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12303)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @NiLuJe)

**Comment by NiLuJe** [comment]
> Turns out the legacy rotation handling that's taking up so much code all over the place is broken, lol.

Well, it sorts of do something completely unhelpful with DjVu, but that's barely any better than mupdf's blank pages ;p.

**Comment by Frenzie** [comment]
Legacy rotation handling?

**Comment by NiLuJe** [comment]
I'm assuming its somewhat legacy, both because it's absolutely not how we handle rotation (which is at blit time via blitbuffer), and by the fact that it's horribly broken ;D.

**Comment by NiLuJe** [comment]
> I'm assuming its somewhat legacy

Especially since the djvu.c comment about that, which already hints at the fact that "this doesn't really make sense anymore" mentions djvureader.lua, which sort of takes us back ;p.

**Comment by Commodore64user** [comment]
> I'm assuming its somewhat legacy, both because it's absolutely not how we handle rotation (which is at blit time via blitbuffer), and by the fact that it's horribly broken ;D.
> 
> That, and that tiny ReaderRotation module is the only thing that attempts to leverage it, via key bindings, which doesn't scream shiny and new ;p.

And in case it has not been said enough times, it does not work... for posterity the opening chapter of #12162 has an ample debate regarding the J/K rascals.  ;).

**Comment by Frenzie** [comment]
If they don't work they should be fixed properly, not randomly removed without consensus.

**Comment by NiLuJe** [comment]
I suspect they *never* did, and more importantly: they do not serve *any* purpose, we already handle rotation just fine via BB.

**Comment by NiLuJe** [comment]
By which I mean, if someone is, for some strange reason, attached to this obscure pair of bindings nobody's ever used (cuz nobody noticed in 14 years it was broken), it's just a matter of binding them to IterateRotation true/false instead.

**Comment by NiLuJe** [comment]
(I wholly plan on purging this mess from the codebase in a followup PR, because it's crapping over *a lot* of code for no good reason).

**Comment by Frenzie** [comment]
I'm fine with removing them. I'm trying to convey that I don't like such a removal trying to sneak in under the radar in a PR about hiding the virtual keyboard when it should be its own thing.

**Comment by NiLuJe** [comment]
FWIW, it came about as a by-product of testing the cache hashes, and wondering why the hell rotation was *always* 0 ;).

**Comment by NiLuJe** [comment]
I can keep the comments and punt the actual commenting out of ReaderRotation in a separate PR if you prefer.

**Comment by Frenzie** [comment]
It was really only in reply to
> for posterity the opening chapter of https://github.com/koreader/koreader/pull/12162 has an ample debate regarding the J/K rascals. ;).

The reason it drags on for so long, exhausting everyone (the author included I'm sure) is because of recurring scope expansions.

**Comment by Commodore64user** [comment]
My apologies.

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 1 of 1 files at r11, 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12303)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @NiLuJe)

**Review by Frenzie — APPROVED** [review]

**Comment by fbriere** [comment]
@NiLuJe: I've been testing this over the past few days, and it's working great for me so far.  Thank you for your hard work!

One small observation: the "Original size" button kinda becomes irrelevant, to the point where one could question its very purpose for existing.  Of course, I understand that imageviewer is larger than panel-zoom, but for the average user, we now have a button that toggles between two almost-identical images for every panel (at least when not rotated).

(I think this is made more striking by the contrast with the old behavior, where this button definitely did something.  Not something useful, or even remotely true to the "Original size" label, but it *was* something.  :grin:)

**Comment by NiLuJe** [comment]
Yup, makes sense. That's just a standard ImageViewer, so the buttons are kind of an all-or-nothing deal ;).

File: frontend/document/document.lua [hunk 1]
```diff
@@ -401,17 +407,37 @@ end
...
+    if is_prescaled then
+        hash = self:getPagePartHash(pageno, zoom, rotation, gamma, rect)
+
+        tile = DocCache:check(hash, TileCacheItem)
+    else
+        hash = self:getFullPageHash(pageno, zoom, rotation, gamma)
+
+        tile = DocCache:check(hash, TileCacheItem)
+
+        -- In the is_prescaled branch above, we're *already* only rendering part of the page
```
Comment (poire-z) [inline frontend/document/document.lua:436]: (These random and uneeded blank lines makes this odd to read :))

Comment (NiLuJe) [reply frontend/document/document.lua:436]: Yeah, we usually have differing opinions on that subject ;).

FWIW, the intent here was to make the distinction between the two different `getPageHash` calls clearer despite their similar names.

**Pull request merged**